Feature/project manager

Steffengreiner opened the pull request:
**Purpose of this PR**
This PR introduces the `ProjectManager` dto as an extension of the abstract `Person` class. 
It's intended to store the information of a `project manager` whose e.g. associated with an Offer. 
Additionally it removes the redundant `persontype` property from the `Person `class and it's implementation in the `customer` dto. 
**Misc**
I'm unsure, if we should set a default value for the projectManagers `affiliations` and `AcademicTitle` property or how they should be handled.

Commit:
Implement Project Manager as an extension of the abstract Person Class and remove personType from the Person and Customer Class

Commit:
Revert wrong test implementation in CustomerSpec -.-

Commit:
Update Offer class and Offer Testing with Project Manager dto instead of Person

KochTobi reviewed (commented):
LGTM

Only 1 question: Since we have a DTO here shouldn't it be fine for the values to be null?

Steffengreiner commented:
@KochTobi Good question, the `requireNonNull` property for the properties of the abstract `Person `class was already set before. I'd have no issue removing it since as of now it's only used in the TOMATO Context where we make sure that null values are handled properly.

sven1103 commented on src/main/groovy/life/qbic/datamodel/dtos/general/Person.groovy:
```diff
@@ -46,13 +41,11 @@ abstract class Person {
 
-  Person(String personType,
-         String firstName,
+  Person(String firstName,
```
This change breaks all implementations of the Customer class. 

It would be preferable to mark it as @deprecated and introduce a Builder without the type instead

KochTobi commented on src/test/groovy/life/qbic/datamodel/dtos/business/ProjectManagerSpec.groovy:
```diff
@@ -0,0 +1,43 @@
+ *
+ * @since 1.12.0
+ */
+class ProjectManagerSpec extends Specification {
+
+        def "missing first name shall raise a NullPointerException"() {
+            when:
+            def projectManager = new ProjectManager(null, "", AcademicTitle.NONE, "a.b@c.de", [])
+
+            then:
+            thrown(NullPointerException)
+        }
+
+        def "missing last name shall raise a NullPointerException"() {
+            when:
+            def projectManager = new ProjectManager("", null, AcademicTitle.NONE, "a.b@c.de", [])
+
+            then:
+            thrown(NullPointerException)
+        }
+
+        def "missing email name shall raise a NullPointerException"() {
+            when:
+            def projectManager = new ProjectManager("", "", AcademicTitle.NONE, null, [])
+
+            then:
+            thrown(NullPointerException)
+        }
+
+        def "missing affiliation list shall cause the projectManager to contain an empty list"() {
+            when:
+            def projectManager = new ProjectManager("", "", AcademicTitle.NONE, "a.b@c.de", null)
+            then:
+            projectManager.affiliations != null
+            projectManager.affiliations.isEmpty()
+        }
+    }
```
I think the tests need to be adjusted or removed since this is no longer expected behavior.

KochTobi reviewed (dismissed):
LGTM I would remove the tests that do not test for expected behaviour (NullPointerException)

Commit:
Merge branch 'development' into feature/ProjectManager

sven1103 committed:
Refactor/persons (#89)

This PR introduces some simple builder for the abstract class
Person and the implementations CommonPerson, Customer and
ProjectManager.

Co-authored-by: Steffengreiner <Steffen.Greiner@gmx.de>

Commit:
Add ProjectManagerTest and remove author from CustomerTest

sven1103 committed:
Merge branch 'development' into feature/ProjectManager
